protostar/commands/test/test_collector.py

```diff
@@ -151,3 +170,156 @@ def _preprocess_contract(self, file_path: Path) -> StarknetPreprocessedProgram:
+                test_suite_filepath,
+                TargetTestCasesInfo(test_case_globs=set(), ignored_test_cases=set()),
+            )
+            target_test_cases_info.test_case_globs = filtered_test_cases_mapping[
+                test_suite_filepath
+            ]
+            if test_suite_filepath in ignored_test_cases_mapping:
+                target_test_cases_info.ignored_test_cases = ignored_test_cases_mapping[
+                    test_suite_filepath
+                ]
+
+        return result
+
+    def _find_test_suite_filepaths_from_glob(self, test_suite_glob: str) -> Set[str]:
+        results: Set[str] = set()
+        matches = glob(test_suite_glob, recursive=True)
+        for match in matches:
+            path = Path(match)
+            if path.is_dir():
+                results.update(self._find_test_suite_filepaths_in_dir(path))
+            elif path.is_file() and TestCollector.is_test_suite(path.name):
+                results.add(match)
+        return results
+
+    # pylint: disable=no-self-use
+    def _find_test_suite_filepaths_in_dir(self, path: Path) -> Set[str]:
+        filepaths = set(glob(f"{path}/**/*.cairo", recursive=True))
+        results: Set[str] = set()
+        for filepath in filepaths:
+            if TestCollector.is_test_suite(Path(filepath).name):
+                results.add(filepath)
+        return results
+
+    def _build_test_suites_from_target_test_cases_info_mapping(
+        self,
+        target_test_cases_info_mapping: TargetTestCasesInfoMapping,
+    ) -> List[TestSuite]:
+        test_suites: List[TestSuite] = []
+        for (
+            test_suite_filepath,
+            target_test_cases_info,
+        ) in target_test_cases_info_mapping.items():
+            test_suites.append(
+                self._build_test_suite_from_test_cases_info(
+                    Path(test_suite_filepath),
+                    target_test_cases_info,
+                )
+            )
+        return test_suites
+
+    def _build_test_suite_from_test_cases_info(
+        self,
+        test_suite_path: Path,
+        test_cases_info: TargetTestCasesInfo,
+    ) -> TestSuite:
+        preprocessed = self._preprocess_contract(test_suite_path)
+        collected_test_case_names = set(self._collect_test_case_names(preprocessed))
+
+        # gather test cases that match any test case glob
+        target_test_case_names: Set[str] = set()
+        for test_case_name in collected_test_case_names:
+            if len(test_cases_info.test_case_globs) == 0:
+                target_test_case_names = collected_test_case_names.copy()
+            else:
+                for test_case_glob in test_cases_info.test_case_globs:
+                    if fnmatch(test_case_name, test_case_glob):
+                        target_test_case_names.add(test_case_name)
+
+        # filter out test cases that match any test case glob
+        not_ignored_target_test_case_names = target_test_case_names.copy()
```

Comment:
Why we have to copy?

Reply:
`len(target_test_case_names)` changed during loop execution